core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java

```diff
@@ -0,0 +1,140 @@
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.common.settings.Settings.settingsBuilder;
+import static org.hamcrest.Matchers.equalTo;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+@ESIntegTestCase.SuppressLocalMode
+public class IndexingMasterFailoverIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        final HashSet<Class<? extends Plugin>> classes = new HashSet<>(super.nodePlugins());
+        classes.add(MockTransportService.TestPlugin.class);
+        return classes;
+    }
+
+    public void testMasterFailoverDuringIndexingWithMappingChanges() throws Throwable {
+        logger.info("--> start 4 nodes, 3 master, 1 data");
+        Settings sharedSettings = settingsBuilder()
+                .put(ElectMasterService.DISCOVERY_ZEN_MINIMUM_MASTER_NODES, 2)
+                .put(FaultDetection.SETTING_PING_TIMEOUT, "100ms") // speed things up
+                .put(ZenDiscovery.SETTING_PING_TIMEOUT, "200ms")
+                .put(DiscoverySettings.COMMIT_TIMEOUT, "100ms") // speed things up
+                .build();
+
+        for (int i = 0; i < 3; i++) {
+            internalCluster().startMasterOnlyNode(sharedSettings);
+        }
+
+        String dataNode = internalCluster().startDataOnlyNode(sharedSettings);
+
+        logger.info("--> wait for all nodes to join the cluster");
+
+        Client client = client();
+
+        assertThat(client.admin().cluster().prepareHealth().setWaitForNodes("4").get().isTimedOut(), equalTo(false));
+
+        // We index data with mapping changes into cluster and have master failover at same time
+        client(dataNode).prepareIndex("myindex", "mytype").setSource("field1", "val1").get();
+
+        final BlockingQueue<Throwable> exceptions = new LinkedBlockingQueue();
+        Thread[] threads = new Thread[6];
+        final CyclicBarrier barrier = new CyclicBarrier(threads.length);
+
+        for (int j = 0; j < threads.length; j++) {
+            final int threadId = j;
+            threads[j] = new Thread(new Runnable() {
+                @SuppressWarnings("unchecked")
+                @Override
+                public void run() {
+                    try {
+                        barrier.await();
+
+                        for (int i = 0; i < 50; i++) {
+                            // index data with mapping changes
+                            IndexResponse response = client(dataNode).prepareIndex("myindex", "mytype").setSource("field_" + threadId + "_" + i, "val").get();
+                            assertThat(response.isCreated(), equalTo(true));
+                        }
+                    } catch (Throwable t) {
+                        exceptions.offer(t);
```

Comment:
we can throw exceptions on the background thread and the test should fail with uncaught exceptions on a thread.